Fix db engine check in Experiment [Resolves #538]

thcrock opened the pull request:
The experiment checks whether or not the passed-in engine is a base sqlalchemy engine, and converts it into the serializable triage engine if so.

However, this check was broken. This PR fixes it to actually duck-type,
to see if the engine can be pickled to a memory file

thcrock commented on src/triage/experiments/base.py:
```diff
@@ -84,7 +84,7 @@ def __init__(
         self._check_config_version(config)
         self.config = config
 
-        if isinstance(db_engine, Engine):
```
One thing that came to mind is that maybe this check should be in multicore. There shouldn't be any harm in converting this for SingleThreaded, but just for the sake of the check being closer to the reason for the check (e.g. multiprocess code). What do you think?

thcrock replied:
If we do so, I tried to see if there's an even better check, for instance trying to get the pickler directly from the multiprocessing library we're using....but I don't see that exposed at all. We could go full duck-type check and actually fork a process with the engine and see if it works.

jesteria replied:
Yeah I isolated the multiprocessing pickler somewhere at some point ... (And, I like the idea of "duck"-typing.)

I think this would be it:

```
from multiprocessing.reduction import ForkingPickler

ForkingPickler.dumps(…)
```

And, regardless, I agree that this logic should be closer to the multiprocessing code, (in `multicore`).

jesteria replied:
That said, I don't love the extra, eager pickling -- bear in mind, rather than duck-typing, we could simply say:

    if isinstance(db_engine, SerializableDbEngine):
        …
    else:
        …

And, eagerly spawning a test fork really seems like overkill.

In fact, **if** it's possible to reliably detect a child process / multiprocessing issue due to serialization, then why check eagerly at all? Could we just let the experiment go, but detect the serialization error and exit (propagating the error)?

That might be nicer than the route of warning and wrapping. (Especially considering that that wrapping code only carries through the DB URL, and the fact that your engine has been replaced with ours *can* be overlooked.)

thcrock replied:
There's value in raising the error immediately after the user starts running and not minutes (if not hours) later when we first encounter the problem, because with the runtimes of experiments people are likely to start them, let them go, and come back later.

thcrock replied:
I'm curious your problem with duck-typing, especially since in a prior comment you said you liked the idea. Is there a performance concern with eagerly pickling something so small as an engine?

jesteria replied:
Yep, I just meant that if we can ensure that the (multiprocessing) procedure raises pickling errors sufficiently early, then this could be better than picking out the engine for pickling that eagerly.

This is only very slightly a performance consideration – (that's likely *not* a significant implication). But, also, because detecting a troublesome engine may be finicky, and because there could be other serialization issues beyond just the engine.

Duck-typing is good; though, it takes various forms, and often isn't "let's just _do the thing_ early to ensure we can do it again later" – though that's not an unreasonable procedure. (Often it's rather "does it 'quack' like a duck," which means checking for indicators of its functionality; but, which I don't know that we _can_ check for, here.)

And, depending on the situation, it is at least as good, and less finicky, to simply "try first, ask questions (that is catch exceptions) later." And depending on the multiprocessing tool I use, this even seems possible – though, perhaps it's not feasible, here.

jesteria replied:
Anyway, now having been reduced to tears by the `multiprocessing` and `pebble` modules … :man_shrugging: :smile: 

I can get `multiprocessing.Pool.map` to raise eager pickling errors -- and it makes sense to catch and propagate those, through the parent process, to the user.

However, that's not what I get with `pebble.ProcessPool.map`; and, I don't know that there's an profitable path to get that, though maybe you have more insight.

And, if we're rather sure that the other components of our multiprocessing task payloads -- other than the engine that is -- "come from us" and should be serializable, then I don't see great value in checking the payloads at the lower level.

So if that's all the case, we are back here, checking the engine :smile:

thcrock replied:
Referring to the duck typing discussion: I get that sometimes we do want to wait until we need to do the thing before we test it out, but I think in triage experiments our priorities are a bit different. 

Referring to the eager process forking: It's a little bit hard to really test all of the process forking in the context that it will be done in the end: A bunch of different components have engines, as well as other arguments that may or may not be serializable. To go that far we would essentially need to do a dry run, or sample run, of the entire experiment with a tiny subset of data.

This 'sample data run' is something we do want to work towards (per the discussion after my presentation on testing a month ago), but it does seem a bit overboard for a bugfix like this.

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,24 @@
+                "Raw, unserializable SQLAlchemy engine passed. "
+                "URL will be used, other options may be lost in multi-process environments"
+            )
+            db_engine = create_engine(db_engine.url)
```
I'm still unsure it makes sense to just write a warning to the logs, and replace their engine, carrying through only the database URL.

It seems more user-friendly, to me – for something like this anyway – to raise an exception, so they can resolve it themselves, (most likely by availing themselves of our `create_engine`), and thereby ensure that their engine is properly created with the arguments they want.

jesteria replied:
Yeah, that's true

jesteria commented on src/triage/util/pickling.py:
```diff
@@ -0,0 +1,23 @@
+        if "pickle" in str(e):
+            return False
+        else:
+            raise
```
I really appreciate this attempt to differentiate pickling errors from others but, especially considering how finicky it is to positively identify pickling errors, and considering the extreme unlikeliness that an exception here is due to anything *but* that the object cannot be pickled, why not just test that any exception is raised?

We could, instead, use the ForkingPickler, and get even closer to what the multiprocessing module tries to do. Regardless, if what it wants to do, (pickle), raises an exception, we know that "it can't be pickled."

jesteria commented on src/triage/util/pickling.py:
```diff
@@ -0,0 +1,23 @@
+    """Check whether or not an object can be pickled by attempting to pickle it to
+    a BytesIO
+
+    Since it pickles to an in-memory object, be wary about using on huge objects!
```
This last bit is true, but it's just a `str` – or rather, in Py3k lingo – a `bytes`, not a `BytesIO`, right?

jesteria replied:
Anyway that seems like a kind of technical implementation detail, so perhaps this could just be a short-and-sweet docstring:

```
Return whether or not an object can be pickled.

Attempts to pickle the given argument in memory via the `pickle` module.
Since it pickles to memory, be wary about using it on huge objects!
```

thcrock replied:
Sure.  However given the other comment, if I'm using the FOrkingPickler directly, I probably don't want to create this new module....I'm not generically testing pickling, I'm testing the multiprocessing Pickler. I should just move this try block into MultiCore directly. Sound right?

jesteria replied:
Sure, sounds good to me

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,25 @@
 class MultiCoreExperiment(ExperimentBase):
-    def __init__(self, n_processes=1, n_db_processes=1, *args, **kwargs):
-        super(MultiCoreExperiment, self).__init__(*args, **kwargs)
+    def __init__(self, db_engine, n_processes=1, n_db_processes=1, *args, **kwargs):
```
it looks like this changes the signature established by the base class in a slightly mean way. perhaps best to say:

```suggestion
    def __init__(self, config, db_engine, *args, n_processes=1, n_db_processes=1, **kwargs):
```

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,25 @@
+                "(e.g. from triage import create_engine)"
+            ) from e
+
+        super(MultiCoreExperiment, self).__init__(db_engine=db_engine, *args, **kwargs)
```
and of course then here:

```suggestion
        super(MultiCoreExperiment, self).__init__(config, db_engine, *args, **kwargs)
```

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,25 @@
+    def __init__(self, db_engine, n_processes=1, n_db_processes=1, *args, **kwargs):
+        try:
+            ForkingPickler.dumps(db_engine)
+        except Exception as e:
```
at least three letters of variable might be nice :smile: 
```suggestion
        except Exception as exc:
```

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,25 @@
+                "multiprocessing is unable to pickle passed SQLAlchemy engine."
+                "use triage.create_engine instead when running MultiCoreExperiment: "
+                "(e.g. from triage import create_engine)"
+            ) from e
```
great use of `from`

jesteria commented on src/triage/experiments/multicore.py:
```diff
@@ -2,15 +2,25 @@
+            ForkingPickler.dumps(db_engine)
+        except Exception as e:
+            raise ValueError(
+                "multiprocessing is unable to pickle passed SQLAlchemy engine."
```
i think you're missing a space before your new sentence :smile: 
```suggestion
                "multiprocessing is unable to pickle passed SQLAlchemy engine. "
```

thcrock commented on src/tests/test_experiments.py:
```diff
@@ -480,3 +481,17 @@ def test_baselines_with_missing_features(experiment_class):
+    with testing.postgresql.Postgresql() as postgresql:
+        db_engine = sqlalchemy.create_engine(postgresql.url())
+        with TemporaryDirectory() as temp_dir:
+            with pytest.raises(ValueError):
```
Yeah makes sense. My internal error decision tree is basically "Can maybe interpret as a ValueError? If yes, ValueError"

But TypeError makes more sense here

thcrock commented:
Made the changes from jesse's review. Since the assignee isn't here until the middle of next week, and I think I've addressed all the concerns that they brought up, and the bug this fixes is a rather onerous one for current users of Triage, I'm opting to just merge this.
